running-code/execute-package.md

```diff
@@ -0,0 +1,163 @@
+be passed a file path for execution. Alternatively you can also pass the name of a module, exactly as would be used after an `import`.
+In this case, Python will look up the module referenced in the
+[currently active environment](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#create-and-use-virtual-environments),
+and will execute it as a script.
+
+This execution mode is performed with the `-m` flag, as in `python -m site`. It can be used in place of a file
+path, but cannot be used in combination with a path, as there can only be one executing module.
+
+:::{tip}
+These commands both do the same thing, but the latter is easier to remember and much more portable, as it doesn't rely
+on the local machine's file system details.
+
+```bash
+python ./.venv/lib/python3.12/site-packages/pip/__main__.py
+```
+
+```bash
+python -m pip
+```
+:::
+
+### Further exploration
+
+On your own or in small groups:
+
+* Install the `my_program.py` module from the [last lesson](execute-script-launch-command)
+* Try to get the same greeting as before using `python -m my_program`.
+
+**my_program.py** which prints out `✨ Hello from Python ✨` when executed as a script.
+
+```python
+#!/usr/bin/env python
+# The above line is a shebang and can take the place of typing python on the command line
+# This comment is below, because shebangs must be the first line!
+
+def shiny_hello():
+    print("\N{Sparkles} Hello from Python \N{Sparkles}")
+
+if __name__ == "__main__":
+    shiny_hello()
+```
+
+### Executable packages
+
+The `-m` flag as described above only works for Python modules (files), but does not work for Python (sub-)packages (directories).
+This means that you cannot execute a command using only the name of your package when it is structured to use directories
+
+Once your package grows, the top-level name `my_program` turns into a directory.
+(See [Python Package Structure](https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-structure.html)
+for when and how to create a package structure).
+```
+project/
+└── src/
+    └── my_program/
+        ├── __init__.py
+        └── greeting.py
+```
+
+However, you can never execute a directory.
+```bash
+python -m my_program
+python: No module named my_program.__main__; 'my_program' is a package and cannot be directly executed
+```
+
+Initially, Python seems to tell you that the directory names, including your top-level package name,
+cannot be directly executed. But the error message contains the hint that you need to make this run properly.
+
+[Earlier you learned](execute-script-name-eq-main) that `if __name__ == "__main__":` can protect parts of your
+script from executing when it is imported, making that conditional change the file's behavior when used as a script.
+There is a very similar concept that can be applied to Python packages.
```

Comment:
I think there's an opportunity here to stitch this in with previous concepts - 

> Remember how our `__init__.py` file is a special file that tells python that our directory is a python package? There is another special file `__main__.py` that allows us to execute that directory! Earlier you learned [...existing text...] - the `__main__.py` file is a way of separating that code into a separate file rather than putting it at the bottom of a module.

Then I think it would be really nice to extend the previous example so that we preserve the `python -m my_program.greeting` call and show how you can do the same thing in multiple ways. So as above in (current) lines 73-82 where we're talking about how projects can expand or get more complicated over time, we could say something like "say your greeting function got very complicated and you needed to split it out into several files like this

```
project/
└── src/
    └── my_program/
        ├── __init__.py
        └── greeting/
            ├── __init__.py
            ├── complicated_stuff.py
            ├── ...
```

You can preserve the ability to call your greeting with `python -m my_program.greeting` with a `__main__.py` file ...

Reply:
I took a stab at folding in the idea of `__init__.py` into this section. LMKWYT